Add brew release command

Rylan12 opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?
- [ ] Have you successfully run `brew man` locally and committed any changes?

-----

This PR adds a new `brew release` developer command to automate the release process. The command, by default, will simply bump the patch version number. If `--major` or `--minor` is passed, the command will bump the major and minor version numbers respectively (and reset other numbers to 0 as needed).

This PR is still _extremely_ rough (hence being marked as a draft). I'm going to immediately cancel CI because I know there are issues that I haven't yet fixed locally. However, I have a few general questions, so I figured I'd open this to track my progress.

I also plan on updating the [Release docs](https://docs.brew.sh/Releases) in this PR in the future.

Additionally, there's some overlap between this command and `brew release-notes` that I plan on extracting to a new module.

-----

## Questions

1. Should there be any interactivity in this command? At the moment, the whole thing runs in about 5 seconds so there isn't really any time for a frantic control-c if someone realizes they messed up. I wonder if we should make this require an additional confirmation before pushing the tags and creating the release. I know we generally like to avoid interactivity, but I don't think this command should even be used in a script so I'm not super worried about that. I'm open to other options here as well.
2. What is the best way to determine the "current tag." At the moment, I'm simply running a `git fetch origin` (although I'm now realizing I probably want `git fetch --tags origin`) before the command starts and then simply taking the highest version number from `git tag --list --sort=-version:refname *.*.*`. Is there a way to determine the latest _upstream_ tag rather than relying on local tags? I guess an alternative would be to query GitHub for the latest release and then use that as the base. An extension question would be: is it okay to assume that those who will run this command will be in a reasonable repository state?
3. How should non-tip-of-the-master-branch releases be handled? I know that there's debate about this in #9331 so I don't need to duplicate that here. For the moment, I think I'm going to assume the following: the `brew release` command will be designed for creating releases _only_ at the tip of the master branch. If a non-tip-of-the-master-branch release is desired, this command should not be used. Is that acceptable? If so, I'll probably either call `brew update` within the command or simply fail if the brew repo isn't up to date (or, to be more precise, I'll check that `git rev-parse HEAD` matched `git rev-parse origin/HEAD`).
4. I think we should sign tags when possible. Should we run `git tag --sign` by default? If not, should we add a `--sign` flag to `brew release` to control this? I have `tag.gpgSign` enabled in my `~/.gitconfig` so my tags would be signed either way, but I'm sure tha not all maintainers do.
5. More specifically, how do I handle the scopes for the GitHub API call? I'm a little confused with what needs to be specified here. [`GitHub.open_api`](https://github.com/Homebrew/brew/blob/7242cd64743635d13d5f69fa3ae0718b2170e7da/Library/Homebrew/utils/github.rb#L179-L243) takes an optional `scopes` argument. In [`GitHub.create_pull_request`](https://github.com/Homebrew/brew/blob/7242cd64743635d13d5f69fa3ae0718b2170e7da/Library/Homebrew/utils/github.rb#L393-L398), the [`GitHub.CREATE_ISSUE_FORK_OR_PR_SCOPES`](https://github.com/Homebrew/brew/blob/7242cd64743635d13d5f69fa3ae0718b2170e7da/Library/Homebrew/utils/github.rb#L22) constant which is set to `public_repo` is passed. What (if any) scope needs to be used for creating a release? Obviously, some write access to the repo is needed. Just not totally sure what to do here and hoping someone can guide me through it :)

-----

## Quick Summary

Based on my answers to the questions above, here is what I'm picturing (at the moment) will happen when this command is run:

1. `git fetch --tags origin` is run in the background.
2. If `git rev-parse HEAD` doesn't match `git rev-parse origin/HEAD`, fail
3. Find the most recent tag using `git tag --list --sort=-version:refname *.*.*`
4. Generate the new tag based on `args.major?` and `args.minor?`
5. If creating a major/minor release, check to make sure the latest release was less than a month ago. If so, fail unless `--force` is passed (which should change the failure into a warning that's still displayed)
5. Create the new tag locally using `git tag` and maybe `--sign`
6. Generate the release notes
7. Asking for a final confirmation before pushing the tag and creating the release
8. Run `git push origin <tag>` to push the tag
9. Create a new release using the GitHub API
10. Print the url to the release
10. Open the url in a browser unless `--no-browse` is passed.

CC @MikeMcQuaid

dawidd6 commented on Library/Homebrew/utils/github.rb:
```diff
@@ -397,6 +398,14 @@ def create_pull_request(repo, title, head, base, body)
     open_api(url, data: data, scopes: scopes)
   end
 
+  def create_release(tag_name:, name:, body:, args:)
```
Look below for `create_or_update_release` method. I think you can just slightly modify it for this command needs instead of making a new method.

Rylan12 replied:
Oh, thanks! Didn't even check for some reason 😅 I'll try to adapt that if possible

MikeMcQuaid commented:
> 1. Should there be any interactivity in this command?

No.

> 1. I wonder if we should make this require an additional confirmation before pushing the tags and creating the release.

I would say it's not worth pushing the tags (which cannot be undone) but instead telling the maintainer what command to run to do so.

> 2\. What is the best way to determine the "current tag." At the moment, I'm simply running a `git fetch origin` (although I'm now realizing I probably want `git fetch --tags origin`) before the command starts and then simply taking the highest version number from `git tag --list --sort=-version:refname *.*.*`.

Sounds good.

> 2\. Is there a way to determine the latest _upstream_ tag rather than relying on local tags? I guess an alternative would be to query GitHub for the latest release and then use that as the base.

Yes, query GitHub.

> 2\. An extension question would be: is it okay to assume that those who will run this command will be in a reasonable repository state?

Yes.

> 3\. For the moment, I think I'm going to assume the following: the `brew release` command will be designed for creating releases _only_ at the tip of the master branch.

Sounds good (forever, in my opinion).

> 3\. If a non-tip-of-the-master-branch release is desired, this command should not be used. Is that acceptable?

Yes.

> 4\. I think we should sign tags when possible. Should we run `git tag --sign` by default?

Yes.

> 5\. [`GitHub.open_api`](https://github.com/Homebrew/brew/blob/7242cd64743635d13d5f69fa3ae0718b2170e7da/Library/Homebrew/utils/github.rb#L179-L243) takes an optional `scopes` argument. In [`GitHub.create_pull_request`](https://github.com/Homebrew/brew/blob/7242cd64743635d13d5f69fa3ae0718b2170e7da/Library/Homebrew/utils/github.rb#L393-L398), the [`GitHub.CREATE_ISSUE_FORK_OR_PR_SCOPES`](https://github.com/Homebrew/brew/blob/7242cd64743635d13d5f69fa3ae0718b2170e7da/Library/Homebrew/utils/github.rb#L22) constant which is set to `public_repo` is passed. What (if any) scope needs to be used for creating a release?

`public_repo` should be sufficient. The GitHub API docs should be able to confirm.

MikeMcQuaid commented on Library/Homebrew/dev-cmd/release.rb:
```diff
@@ -0,0 +1,119 @@
+    Homebrew::CLI::Parser.new do
+      description <<~EOS
+        Automate the Homebrew release process. Running `brew release` will
+        create a new tag, push the tag, and create a new release with release notes.
```
As mentioned: I'd suggest this doesn't push.

MikeMcQuaid commented on Library/Homebrew/dev-cmd/release.rb:
```diff
@@ -0,0 +1,119 @@
+             description: "Create a major release."
+      switch "--minor",
+             description: "Create a minor release."
+      switch "--force-major-minor",
+             description: "Create a release despite the warning that the latest " \
+                          "major/minor release was made less than one month ago."
+      switch "--force-mismatched-head",
+             description: "Create a release despite the warning that HEAD does not match origin/HEAD."
+      switch "--no-push",
+             description: "Create the new tag locally but do not push to the remote."
+      switch "--no-browse",
+             description: "Print the release URL instead of opening in a browser."
```
Let's remove all of these options.

MikeMcQuaid commented on Library/Homebrew/dev-cmd/release.rb:
```diff
@@ -0,0 +1,119 @@
+  def release
+    args = release_args.parse
+
+    safe_system "git", "-C", HOMEBREW_REPOSITORY, "fetch", "origin"
```
Could perhaps instead make this command one that always runs `brew update` before.

Rylan12 replied:
Yeah, I like this better (although then it could be overridden by `HOMEBREW_NO_AUTO_UPDATE`. One question, though. How do I do that? I've never figured out how those commands to that 😅

Rylan12 replied:
Ah, nevermind. This may not even be necessary if we query the GitHub API for the latest tag anyway. I'll need to play around more I think

MikeMcQuaid replied:
Agreed probably no longer necessary but: to answer your question: https://github.com/Homebrew/brew/blob/60325a6f739a3c476804b5acc432c9f64a484279/Library/Homebrew/brew.sh#L162-L177

MikeMcQuaid commented on Library/Homebrew/dev-cmd/release.rb:
```diff
@@ -0,0 +1,119 @@
+    local_head = Utils.popen_read("git", "-C", HOMEBREW_REPOSITORY, "rev-parse", "HEAD").chomp
+    upstream_head = Utils.popen_read("git", "-C", HOMEBREW_REPOSITORY, "rev-parse", "origin/HEAD").chomp
+    if local_head != upstream_head
+      message = "The current HEAD does not match origin/HEAD! Pass --force-mismatched-head to override."
```
I'd say: don't even provide the option.

MikeMcQuaid commented on Library/Homebrew/dev-cmd/release.rb:
```diff
@@ -0,0 +1,119 @@
+      current_version_date = Date.parse Utils.popen_read(
+        "git", "-C", HOMEBREW_REPOSITORY, "log", "-1", "--format=%aI", "#{current_version.major_minor}.0"
+      )
+      message = "The latest major/minor release was less than one month ago. Pass --force-major-minor to override."
```
Same here: let's error but not provide the option.

MikeMcQuaid commented on Library/Homebrew/dev-cmd/release.rb:
```diff
@@ -0,0 +1,119 @@
+    end
+
+    begin
+      release = GitHub.create_release tag_name: new_version, name: new_version, body: release_notes.join("\n"), args: args
```
Could consider creating just a draft release, not pushing the tag and allowing the user to use the GitHub UI to publish the release (which will create the tag).

MikeMcQuaid reviewed (commented):
Feel free to punt until PR further along but: would also be nice to reference this in the release docs when this PR is shipped.

Rylan12 commented:
Thanks, that's good feedback. I'm in agreement about not pushing the tags. Makes sense to not even risk it.

Can you create a draft release that isn't attached to a tag? I'll have to play around with it. It's certainly ideal if that's the case.

> > 5. [`GitHub.open_api`](https://github.com/Homebrew/brew/blob/7242cd64743635d13d5f69fa3ae0718b2170e7da/Library/Homebrew/utils/github.rb#L179-L243) takes an optional `scopes` argument. In [`GitHub.create_pull_request`](https://github.com/Homebrew/brew/blob/7242cd64743635d13d5f69fa3ae0718b2170e7da/Library/Homebrew/utils/github.rb#L393-L398), the [`GitHub.CREATE_ISSUE_FORK_OR_PR_SCOPES`](https://github.com/Homebrew/brew/blob/7242cd64743635d13d5f69fa3ae0718b2170e7da/Library/Homebrew/utils/github.rb#L22) constant which is set to `public_repo` is passed. What (if any) scope needs to be used for creating a release?
> 
> `public_repo` should be sufficient. The GitHub API docs should be able to confirm.

This is saying that the token being used must have access to all `public_repo`s that the user it's associated with has access to, right?

MikeMcQuaid commented:
> Can you create a draft release that isn't attached to a tag?

Yup! The tag isn't created until it's published. That's often the flow I do for creating a release.



> This is saying that the token being used must have access to all `public_repo`s that the user it's associated with has access to, right?

Yes, think so.

Rylan12 commented:
Okay, here's where this command currently stands.

Running `brew release` will create a draft PR with the correct release notes. It doesn't actually create the tag locally and push but relies on the tag being created when the new release is published. By default, the tag is created at the tip of the `master` branch. I believe this can be customized if we want, instead, to create the tag at the `HEAD` commit of the user running the release command. Since in almost all cases the release it at the tip (I know there's a debate about this), I don't think this is necessary.

The `--major` and `--minor` flags can be passed to create a new major or minor release, respectively. If neither is passed, a patch release will be made. The command will fail if a major/minor release is requested and the latest major/minor release was published less than one month ago (determined by querying the GitHub API).

In order to generate the appropriate release notes, this command needs to have the latest commits on `origin/master` fetched locally. I've added the auto-update functionality to the `brew release` command. To work around the case where `HOMEBREW_NO_AUTO_UPDATE` is set (which may be true for maintainers), I added an extra step that runs `git fetch origin` _only_ if `HOMEBREW_NO_AUTO_UPDATE` is set. This will ensure that the latest master branch commits are available locally. **Do we need to explicitly call `git fetch origin master` or is `git fetch origin` enough? Or another command?**

Alternatively, another option is to query the GitHub API for the latest commit on the `master` branch. If it detects that the local `origin/HEAD` does not point to the same commit that the GitHub API reports, it will prompt the user to run `brew update`. Not sure which is preferred.

I've also refactored some of the shared logic between the `brew release` command and the `brew release-notes` command. I think the two commands are distinct enough that they should both be kept. However, the `brew release-notes` command is no longer used in our release process so could theoretically be removed.

Lastly, I've updated the release documentation.

SeekingMeaning commented on Library/Homebrew/release_notes.rb:
```diff
@@ -0,0 +1,35 @@
+  module_function
+
+  sig {
+    params(start_ref: T.any(String, Version), end_ref: T.any(String, Version), markdown: T::Boolean)
```
```suggestion
    params(start_ref: T.any(String, Version), end_ref: T.any(String, Version), markdown: T.nilable(T::Boolean))
```

SeekingMeaning commented on Library/Homebrew/dev-cmd/release-notes.rb:
```diff
@@ -55,25 +56,9 @@ def release_notes
-        "- [#{Regexp.last_match(3)}](#{Regexp.last_match(1)}) (@#{Regexp.last_match(2)})"
-      end
-    end
+    release_notes = ReleaseNotes.generate_release_notes previous_tag, end_ref, markdown: T.must(args.markdown?)
```
```suggestion
    release_notes = ReleaseNotes.generate_release_notes previous_tag, end_ref, markdown: args.markdown?
```

Rylan12 replied:
Why is this way preferred? The way I see it, `generate_release_notes` should only accept `T::Boolean` for the `markdown` parameter.

It's up to whatever is calling the method to ensure that it's passing a boolean. In this case, `T.must` is appropriate because we can be sure that `args.markdown?` is either `true` or `false` because `--markdown` is one of the arg options.

SeekingMeaning replied:
> we can be sure that `args.markdown?` is either `true` or `false` because `--markdown` is one of the arg options.

Correct me if I'm wrong, `args.markdown?` returns `nil` if `--markdown` isn't passed, right?

Rylan12 replied:
Oops, you're right! In that case, I agree with this change.

MikeMcQuaid commented:
> Running `brew release` will create a draft PR with the correct release notes. It doesn't actually create the tag locally and push but relies on the tag being created when the new release is published. By default, the tag is created at the tip of the `master` branch. I believe this can be customized if we want, instead, to create the tag at the `HEAD` commit of the user running the release command. Since in almost all cases the release it at the tip (I know there's a debate about this), I don't think this is necessary.

👍🏻 

> In order to generate the appropriate release notes, this command needs to have the latest commits on `origin/master` fetched locally. I've added the auto-update functionality to the `brew release` command. To work around the case where `HOMEBREW_NO_AUTO_UPDATE` is set (which may be true for maintainers), I added an extra step that runs `git fetch origin` _only_ if `HOMEBREW_NO_AUTO_UPDATE` is set. This will ensure that the latest master branch commits are available locally. **Do we need to explicitly call `git fetch origin master` or is `git fetch origin` enough? Or another command?**
> 
> Alternatively, another option is to query the GitHub API for the latest commit on the `master` branch. If it detects that the local `origin/HEAD` does not point to the same commit that the GitHub API reports, it will prompt the user to run `brew update`. Not sure which is preferred.

I am fine with any of these options. I'd probably suggest "whichever one requires the least code to handle this edge case".

> I've also refactored some of the shared logic between the `brew release` command and the `brew release-notes` command. I think the two commands are distinct enough that they should both be kept. However, the `brew release-notes` command is no longer used in our release process so could theoretically be removed.

Good to know. Perhaps it's worth deprecating? Could add a commented-out `odeprecated` and I can remove on the next release?

MikeMcQuaid commented on Library/Homebrew/dev-cmd/release.rb:
```diff
@@ -0,0 +1,88 @@
+  module_function
+
+  sig { returns(CLI::Parser) }
+  def release_args
```
Could this get a tests for the args (like release-notes has)?

MikeMcQuaid commented on Library/Homebrew/release_notes.rb:
```diff
@@ -0,0 +1,35 @@
+# Helper functions for generating release notes.
+#
+# @api private
+module ReleaseNotes
```
Any unit tests for this module would be nice (but not blocking).

Rylan12 commented:
> > In order to generate the appropriate release notes, this command needs to have the latest commits on `origin/master` fetched locally. I've added the auto-update functionality to the `brew release` command. To work around the case where `HOMEBREW_NO_AUTO_UPDATE` is set (which may be true for maintainers), I added an extra step that runs `git fetch origin` _only_ if `HOMEBREW_NO_AUTO_UPDATE` is set. This will ensure that the latest master branch commits are available locally. **Do we need to explicitly call `git fetch origin master` or is `git fetch origin` enough? Or another command?**
> > Alternatively, another option is to query the GitHub API for the latest commit on the `master` branch. If it detects that the local `origin/HEAD` does not point to the same commit that the GitHub API reports, it will prompt the user to run `brew update`. Not sure which is preferred.
> 
> I am fine with any of these options. I'd probably suggest "whichever one requires the least code to handle this edge case".

Sounds good. I went with the one-line `git fetch` solution.

Rylan12 commented:
Okay, I'm going to go ahead and merge this. @MikeMcQuaid let me know how this works! (I'm assuming you'll be the one to make the next release).
